phoenix-core/src/main/java/org/apache/phoenix/cache/TenantCacheImpl.java

```diff
@@ -77,57 +152,132 @@ public MemoryManager getMemoryManager() {
+    }
+
+    private Cache<ImmutableBytesPtr, CacheEntry> buildCache(final int ttl, final boolean isPersistent) {
+        CacheBuilder<Object, Object> builder = CacheBuilder.newBuilder();
+        if (isPersistent) {
+            builder.expireAfterWrite(ttl, TimeUnit.MILLISECONDS);
+        } else {
+            builder.expireAfterAccess(ttl, TimeUnit.MILLISECONDS);
+        }
+        return builder
+            .ticker(getTicker())
+            .removalListener(new RemovalListener<ImmutableBytesPtr, CacheEntry>(){
+                @Override
+                public void onRemoval(RemovalNotification<ImmutableBytesPtr, CacheEntry> notification) {
+                	if (isPersistent || !notification.getValue().getUsePersistentCache()) {
+                        Closeables.closeAllQuietly(Collections.singletonList(notification.getValue()));
+                	}
+                }
+            })
+            .build();
+    }
     
-    @Override
+    private void evictInactiveEntries(long bytesNeeded) {
+        CacheEntry[] entries = getPersistentServerCaches().asMap().values().toArray(new CacheEntry[]{});
+        Arrays.sort(entries);
+        long available = this.getMemoryManager().getAvailableMemory();
+        for (int i = 0; i < entries.length && available < bytesNeeded; i++) {
+            CacheEntry entry = entries[i];
+            if (!entry.isLive()) {
+            	getServerCaches().invalidate(entry.getCacheId());
+            	getPersistentServerCaches().invalidate(entry.getCacheId());
+                available = this.getMemoryManager().getAvailableMemory();
+            }
+        }
+    }
+
+    private CacheEntry getIfPresent(ImmutableBytesPtr cacheId) {
+        CacheEntry entry = getPersistentServerCaches().getIfPresent(cacheId);
+        if (entry != null) {
+            getServerCaches().put(cacheId, entry);
+        }return getServerCaches().getIfPresent(cacheId);
+    }
+
+	@Override
     public Closeable getServerCache(ImmutableBytesPtr cacheId) {
         getServerCaches().cleanUp();
-        return getServerCaches().getIfPresent(cacheId);
+        CacheEntry entry = getIfPresent(cacheId);
+        if (entry == null) {
+            return null;
+        }
+        return entry.closeable;
     }
-    
+
     @Override
-    public Closeable addServerCache(ImmutableBytesPtr cacheId, ImmutableBytesWritable cachePtr, byte[] txState, ServerCacheFactory cacheFactory, boolean useProtoForIndexMaintainer, int clientVersion) throws SQLException {
+    public Closeable addServerCache(ImmutableBytesPtr cacheId, ImmutableBytesWritable cachePtr, byte[] txState, ServerCacheFactory cacheFactory, boolean useProtoForIndexMaintainer, boolean usePersistentCache, int clientVersion) throws SQLException {
         getServerCaches().cleanUp();
-        MemoryChunk chunk = this.getMemoryManager().allocate(cachePtr.getLength() + txState.length);
+        long available = this.getMemoryManager().getAvailableMemory();
```

Comment:
Two scary points (that I hope I just missed via my scanning):

* I don't see a corresponding update to the GlobalMemoryManager. Are we accounting for the size of `persistentServerCaches` somewhere else?
* Do users who execute queries with persistentCaches have the ability to deny the queries of users not running persistentCaches (by saturating the total RS's serverCache). The 30min default timeout to start evicting persistentCaches makes me worry as that's a long time.

Reply:
I am fairly certain that we export serverCache metrics to the HBase RS metrics subsystem. Assuming that we do, can we break out the persistentServerCache usage as well? That would be nice to monitor.

Reply:
> I don't see a corresponding update to the GlobalMemoryManager. Are we accounting for the size of persistentServerCaches somewhere else?

I think this is handled, but I could be wrong. There is a call like this: `MemoryChunk chunk = this.getMemoryManager().allocate(size);`. This allocates the server cache. The server cache can then live in one or two hash maps, either only `getServerCaches()`, or both `getServerCaches()` and `getPersistentServerCaches()`. However, even if it lives in both, it only uses `size` bytes, since the two maps have a pointer to the same memory chunk.

Reply:
> since the two maps have a pointer to the same memory chunk.

Ah, that's the rub. I assumed that when the persistentCaches was used, there would increased memory consumption over when they were not used. If memory use is constant (or at least, negligible), then I agree with you.